### DISCUSSION TITLE
ComplexMethod diagnostic

### DISCUSSION THREAD

**Discussion by pasniak** [detekt/detekt#3472] [Q&A] [answered]
Is there a way to see some diagnostic explaining why ComplexMethod is triggered? (more precisely why complexity score is what it is) I sometimes get ComplexMethod for functions with local functions and I struggle to see why....

**Comment by schalkms** [comment, marked as answer]
This rule uses McCabe’s Cyclomatic Complexity (MCC) metric to measure the number of linearly independent paths through a function’s source code (https://www.ndepend.com/docs/code-metrics#CC). The higher the number of independent paths, the more complex a method is.

[Source](https://detekt.github.io/detekt/complexity.html#complexmethod)

**Comment by pasniak** [comment]
Thanks for sending the link to `ComplexMethod`'s Configuration options; it lists nesting functions (ignoreNestingFunctions) so how those contribute to the score (in terms of paths)?

Ideally a dump or log showing the paths and how they contribute to a score (ideally I see it as some highlighting of if/while/for... contributing to the score in the plugin when the cursor is in the function name) would be great.

**Reply by BraisGabin** [reply]
> Ideally a dump or log showing the paths and how they contribute to a score (ideally I see it as some highlighting of if/while/for... contributing to the score in the plugin when the cursor is in the function name) would be great.

That sounds like a great improvement for the issue message. Do you have any idea about how that log should look?

**Comment by pasniak** [comment]
Given https://github.com/detekt/detekt/blob/4ce376e282bfff176d9b5b141462e623aab91660/detekt-metrics/src/main/kotlin/io/github/detekt/metrics/CyclomaticComplexity.kt#L85  code doing the CC calculation:

As `visit{Something}` is called back on AST, accumulate strings such as "+1 if" "+n when" (where n is complexity added) etc... in a tree. Dump the (ideally indented...) tree. And print tree as Complexity Report* Detail or supplement. Only run on just edited code. For clarity maybe is separate `class CyclomaticComplexityDetail` 

For starters, just log (println) on each `visit{Something}`? 

*) for example I'd see this in terminal which started Android Studio:
```
Complexity Report:
	- 1 lines of code (loc)
	- 1 source lines of code (sloc)
	
Complexity Report Detail:
fun f()
+2 try/catch
+3  when
```

for some `fun f() = try { when (x) { 1 ->{}; 2->{}; 3->{} } catch (e:Ex1) {} catch (e:Ex2) {}`

**Comment by schalkms** [comment]
For exactly this reason, detekt includes the possibility to extend the feature set by using custom rules and reports.
Please have a look at the following page and give us feedback. Thank you!

https://detekt.github.io/detekt/extensions.html
